gtags/gtags.1
.\" This file is generated automatically by convert.pl from gtags/manual.in.
.TH GTAGS 1 "February 2021" "GNU Project"
.SH NAME
gtags \- create tag files for global
.SH SYNOPSIS
\fBgtags\fP [-ciIOqvw][-C \fIdir\fP][-d \fItag-file\fP][-f \fIfile\fP][\fIdbpath\fP]
.br
.SH DESCRIPTION
\fBGtags\fP is used to create tag files for \fBglobal\fP(1).
.PP
\fBGtags\fP recursively collects source files under the current directory,
picks up symbols and writes the cross-reference data into the tag files
(\'GTAGS\', \'GRTAGS\' and \'GPATH\').
.PP
By default, \fBgtags\fP picks up
C, Yacc, assembly language, Java, C++ and PHP source files.
Files whose names end in \'.c\', \'.h\' are assumed to be C source files.
Files whose names end in \'.y\' are assumed to be Yacc source files.
Files whose names end in \'.s\', \'.S\' are assumed to be assembly language source files.
Files whose names end in \'.java\' are assumed to be Java source files.
Files whose names end in \'.c++\', \'.cc\', \'.hh\', \'.cpp\', \'.cxx\', \'.hxx\', \'.hpp\', \'.C\', \'.H\' are assumed to be C++ source files.
Files whose names end in \'.php\', \'.php3\', \'.phtml\' are assumed to be PHP source files.
Other files are assumed to be text files. Gtags does not treat binary files.
.PP
If \'gtags.files\' exists in the current directory
or a file is specified by the \fB-f\fP option,
target files are limited by it. Lines starting with \'. \' are comments.
.SH OPTIONS
The following options are available:
.PP
.TP
\fB--accept-dotfiles\fP
Accept files and directories whose names begin with a dot.
By default, \fBgtags\fP ignores them.
.TP
\fB-c\fP, \fB--compact\fP
Make \'GTAGS\' in compact format.
This option does not influence \'GRTAGS\',
because that is always made in compact format.
.TP
\fB-C\fP, \fB--directory\fP \fIdir\fP
Change the directory before doing all the work including parameter analysis.
This option is ignored in \fBGTAGS_OPTIONS\fP.
Please specify it on the command line directly.
.TP
\fB--config\fP[=\fIname\fP]
Print the value of config variable \fIname\fP.
If \fIname\fP is not specified then print all names and values.
Additionally, you can refer to the directory name where the system
is installed as read only variables:
\fBbindir\fP, \fBlibdir\fP, \fBdatadir\fP, \fBlocalstatedir\fP and \fBsysconfdir\fP.
.TP
\fB-d\fP, \fB--dump\fP \fItag-file\fP
Dump a tag file as text to the standard output.
Output format is 'key<tab>data'. This is for debugging.
.TP
\fB--explain\fP
Explain handling files.
.TP
\fB-f\fP, \fB--file\fP \fIfile\fP
Give a list of candidates of target files.
Files which are not on the list are ignored.
The argument \fIfile\fP can be set to \'-\' to accept a list of
files from the standard input.
File names must be separated by newline.
To make the list you may use \fBfind\fP(1), which has rich options
for selecting files.
.TP
\fB--gtagsconf\fP \fIfile\fP
Set environment variable \fBGTAGSCONF\fP to \fIfile\fP.
.TP
\fB--gtagslabel\fP \fIlabel\fP
Set environment variable \fBGTAGSLABEL\fP to \fIlabel\fP.
.TP
\fB--help\fP
Print a usage message.
.TP
\fB-I\fP, \fB--idutils\fP
In addition to tag files, make ID database for \fBidutils\fP(1).
.TP
\fB-i\fP, \fB--incremental\fP
Update tag files incrementally.
It's better to use \fBglobal\fP(1) with the \fB-u\fP command.
.TP
\fB-O\fP, \fB--objdir\fP
Use BSD-style obj directory as the location of tag files.
If \fBGTAGSOBJDIRPREFIX\fP is set and \'$GTAGSOBJDIRPREFIX\' directory exists,
\fBgtags\fP creates \'$GTAGSOBJDIRPREFIX/<current directory>\' directory
and makes tag files in it.
Though you can use \fBMAKEOBJDIRPREFIX\fP instead of \fBGTAGSOBJDIRPREFIX\fP,
it is deprecated.
If \fIdbpath\fP is specified, this option is ignored.
.TP
\fB--single-update\fP \fIfile\fP
Update tag files for a single file.
It is considered that \fIfile\fP was added, updated or deleted,
and there is no change in other files.
This option implies the \fB-i\fP option.
.TP
\fB--skip-unreadable\fP
Skip unreadable files.
.TP
\fB--skip-symlink\fP [=\fItype\fP]
Skip symbolic links. If \fItype\fP is 'f' then skip only symbolic links for
file, else if 'd' then skip only symbolic links for directory.
The default value of \fItype\fP is 'a' (all symbolic links).
.TP
\fB--sqlite3\fP
Use Sqlite 3 API to make tag files. By default, BSD/DB 1.85 API is used.
To use this option, you need to invoke configure script with
\fB--with-sqlite3\fP in the build phase.
.TP
\fB--statistics\fP
Print statistics information.
.TP
\fB-q\fP, \fB--quiet\fP
Quiet mode.
.TP
\fB-v\fP, \fB--verbose\fP
Verbose mode.
.TP
\fB--version\fP
Show version number.
.TP
\fB-w\fP, \fB--warning\fP
Print warning messages.
.TP
\fIdbpath\fP
The directory in which tag files are generated.
The default is the current directory.
.SH EXAMPLES
.nf
$ ls -F
Makefile      src/    lib/
$ gtags -v
$ global -x main
main              10 src/main.c  main (argc, argv) {
.fi
.SH FILES
.TP
\'GTAGS\'
Tag file for definitions.
.TP
\'GRTAGS\'
Tag file for references.
.TP
\'GPATH\'
Tag file for source files.
.TP
\'gtags.conf\', \'$HOME/.globalrc\'
Configuration data for GNU Global.
See \fBgtags.conf\fP(5).
.TP
\'gtags.files\'
The list of candidates of target files.
.TP
\'.notfunction\'
The list of symbols which is not a definition.
If this file exists in the project root directory, \fBgtags\fP does not
regard the symbols listed in this file as definitions.
Lines starting with ';' are comment lines.
.PP
.SH ENVIRONMENT
The following environment variables affect the execution of \fBgtags\fP:
.PP
.TP
\fBGTAGSCACHE\fP
The size of the B-tree cache. The default is 50000000 (bytes).
.TP
\fBGTAGSCONF\fP
Configuration file.
.TP
\fBGTAGSFORCECPP\fP
If this variable is set, each file whose suffix is \'.h\' is forcibly
parsed by the built-in C++ parser. Do not set this variable if you are
using plug-in parsers. This variable and the built-in C++ parser are deprecated.
.TP
\fBGTAGSFORCEENDBLOCK\fP
If this variable is set, each } at the first column brings end of block
of 0 level.
.TP
\fBGTAGSLABEL\fP
Configuration label. The default is \'default\'.
.TP
\fBGTAGSLOGGING\fP
If this variable is set, \'$GTAGSLOGGING\' is used as the path name
of a log file. There is no default value.
.TP
\fBGTAGS_COMMANDLINE\fP
This variable can only be referenced from the hook (see gtags_hook).
\fBGtags\fP sets its own effective command line to this variable before
calling the hook. Each argument is separated by whitespace, and
real whitespace is represented as '%20'. This is read only.
.TP
\fBGTAGS_OPTIONS\fP
The value of this variable is inserted in the head of arguments.
.TP
\fBGTAGSOBJDIR\fP
If this variable is set, it is used as the name of BSD-style objdir.
The default is \'obj\'.
Though you can use \fBMAKEOBJDIR\fP instead of \fBGTAGSOBJDIR\fP,
it is deprecated.
.TP
\fBGTAGSOBJDIRPREFIX\fP
If this variable is set, it is used as the prefix of BSD-style objdir.
The default is \'/usr/obj\'.
Though you can use \fBMAKEOBJDIRPREFIX\fP instead of \fBGTAGSOBJDIRPREFIX\fP,
it is deprecated.
.TP
\fBTMPDIR\fP
The location used to stored temporary files. The default is \'/tmp\'.
.SH CONFIGURATION
The following configuration variables affect the execution of \fBgtags\fP.
You can see the default value for each variable with the \fB--config\fP option.
.PP
.TP
gtags_parser (comma separated list)
Specify the mapping of language names and plug-in parsers.
Each part delimited by the comma consists of a language name, a colon,
the shared object path, an optional colon followed by a function name.
If the function name is not specified, 'parser' is assumed.
As a special exception, \fBgtags\fP collects values
from multiple gtags_parser variables.
For these mappings, the first match is adopted.
.TP
gtags_hook (command line)
Specify a command line which should be executed at the beginning of \fBgtags\fP
after loading configuration file. You can use this hook to update
\'gtags.files\' dynamically.
"./" in it always means the project root directory, since \fBgtags\fP is
always invoked there.

.br
This hook is ignored when the following options are specified:
--version, --help, --config, --dump.
.TP
icase_path (boolean)
Ignore case distinctions in the path.
Suffixes check is affected by this capability.
.TP
langmap (comma separated list)
Language mapping. Each comma-separated map consists of
a language name, a colon, and a list of file extensions.
You can specify a glob pattern surrounded by parentheses instead of an extension
for the files without extensions (e.g. Make:([Mm]akefile).mak.mk).
As a special exception, \fBgtags\fP collects values
from multiple langmap variables.
For these mappings, the first match is adopted.
Default mapping is:
.br
\'c:.c.h,yacc:.y,asm:.s.S,java:.java,cpp:.c++.cc.hh.cpp.cxx.hxx.hpp.C.H,php:.php.php3.phtml\'.
.TP
skip (comma separated list)
\fBGtags\fP skips files and directories which are given in this list.
As a special exception, \fBgtags\fP collects values from multiple
skip variables.
If the value ends with \'/\', it is assumed as a directory and
\fBgtags\fP skips all files under it.
The value may include glob patterns (*, ?, [...], [!...], [^...]).

.br
If the value starts with \'/\', it is assumed a relative path name
from the root directory of the project. You cannot use glob patterns
for a path name. However, this direction is out-of-date, and is not
recommended. Instead, you can use \fB-f\fP option which can be combined with
\fBfind\fP(1). Since \fBfind\fP(1) has rich options to select files,
you can do everything. Additionally, this technique can also be applied
to any other tagging systems like \fBctags\fP(1), \fBcscope\fP(1), etc. 

.br
Skip list is also effective when you use the \fB-f\fP or \'gtags.files\'.
.SH EXAMPLES
.nf
#
# There are several ways to use Universal Ctags plug-in parser.
#
$ gtags --gtagsconf=/usr/local/share/gtags/gtags.conf --gtagslabel=universal-ctags
$ gtags --explain
or
$ export GTAGSCONF=/usr/local/share/gtags/gtags.conf
$ export GTAGSLABEL=universal-ctags
$ gtags --explain
or
# In your project
$ cp /usr/local/share/gtags/gtags.conf .
$ vi gtags.conf		# default:tc=native => default:tc=universal-ctags
$ gtags --explain
.fi
.SH DIAGNOSTICS
\fBGtags\fP exits with a non-0 value if an error occurred, 0 otherwise.
.PP
Note that files created by \fBgtags\fP with a non-zero exit code should be
considered corrupted.
.SH "SEE ALSO"
\fBglobal\fP(1),
\fBhtags\fP(1),
\fBgtags.conf\fP(5).
.PP
GNU Global source code tag system
.br
(http://www.gnu.org/software/global/).
.SH BUG
\'GTAGS\' and \'GRTAGS\' are very large.
In advance of using this command, check the space of your disk.
.PP
Assembly language support is far from complete.
It extracts only ENTRY() and ALTENTRY() from source file.
Probably valid only for FreeBSD and Linux kernel source.
.PP
C++ support is deprecated.
.PP
There is no concurrency control about tag files.
.SH AUTHOR
Shigio YAMAGUCHI, Hideki IWAMOTO and others.
.SH HISTORY
The \fBgtags\fP command appeared in FreeBSD 2.2.2.


NEWS
___________________________________
|      |  |  |     |  _  |     |  |
|  |___|  |  |  |  |    _|  |  |  |    GNU GLOBAL source code tagging system
|  |   |  |  |  |  |     |     |  |
|  ~~  |   ~~|     |  ~  |  |  |   ~~|          for all hackers.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 Copyright (c) 2000-2021 Tama Communications Corporation

 This file is free software; as a special exception the author gives
 unlimited permission to copy and/or distribute it, with or without
 modifications, as long as this notice is preserved.

 This program is distributed in the hope that it will be useful, but
 WITHOUT ANY WARRANTY, to the extent permitted by law; without even the
 implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

		----------------------------------

GNU Global NEWS - User visible changes.
At least, you should search '[INCOMPATIBLE CHANGES]' for each version.

Version 6.6.13 - July 2 2024

[FIXED BUGS]
o htags: took the input provided in the dbpath option (-d) and
  appended it to a command string that will later be executed by
  system() function, without any escaping, leading to a command
  injection vulnerability (CVE-2024-38448).

	$ rm -f /tmp/w
	$ htags -g -q -d '$(pwd>/tmp/w)'
	$ cat /tmp/w
	/tmp/global/htags

  Now it just outputs an error:

	$ rm -f /tmp/w
	$ htags -g -q -d '$(pwd>/tmp/w)'
	gtags: directory '/tmp/global/htags/$(pwd>' not found.
	$ cat /tmp/w
	cat: /tmp/w: No such file or directory

o gtags: Pygments plug-in parser with Python3 does not work,
  if 'ctagscom' is not set. If it is not set, default path obtained
  by configure script should be used.
        
	$ cat > gtags.conf
	default:\
		:ctagscom=:\
		:langmap=C\:.c.h:\
		:gtags_parser=C\:/usr/local/lib/gtags/pygments-parser.la:
	$ gtags
	$ global -x '.*'
	$ _                             # no tags

	Now it works.

Version 6.6.12 - March 5 2024

[CHANGES]
global: Added new option 'limit' for the --completion command.
  If limit option is given, print up to the limit lines.
	$ global --completion=100 -P lib

[FIXED BUGS]
o global: The -c command with the --from-here option brings segmentation fault.
	$ global -c --from-here='158:libutil/find.c'
  Now it works (--from-here is ignored).

o htags: If the environment happens to be large, it fails with an obscure
  error message:
	$ htags ...
	htags: Negative exec line limit = -80951
  Now it mostly works since the limit has been loosened.

Version 6.6.11 - November 20 2023

[CHANGE IN SUPPORT STATUS]
Exuberant Ctags as a plug-in parser is no longer supported.
Please use Universal Ctags 6.0.0 or later instead. The same
applies to Pygments plug-in parser.

[FIXED BUGS]
o gtags('pygments-parser.la'): The following setting of Universal Ctags
  caused no tag output. Same result if the value is not 'no'.

        [$HOME/.config/conf.ctags]
        +---------------------------
        |--tag-relative=always

  Now it works.

Version 6.6.10 - May 9 2023

[INCOMPATIBLE CHANGES]
o configure: Auto-detection of python in the config stage is no longer done.
  '$datadir/gtags/script/pygments_parser.py' will have '#!/usr/bin/env python' by default.
  You can specify the path using the following option:
  (Assuming python is located in /usr/local3/bin)

	./configure --with-python-interpreter=/usr/local3/bin/python

o configure: The --with-pread-pwrite option no longer exists.
  Now configure automatically uses pread(2) and pwrite(2) if they are available.

[FIXED BUGS]
o gtags(C++ parser): did not recognize 'namespace name1::name2::name3' syntax.
  Now it picks up 'name2' and 'name3' in addition to 'name1' in the next example:

	namespace name1::name2::name3 { /*...*/ }

o gtags('universal-ctags.la'): The following setting of Universal Ctags caused no tag output.
  Same result if the value is not 'no'.

	[$HOME/.config/conf.ctags]
	+---------------------------
	|--tag-relative=always

  Now it works. This fix applies to 'exuberant-ctags.la' as well.

Version 6.6.9 - December 12 2022

[FIXED BUGS]
o gtags(C++ parser): did not recognize 'final' keyword. Now it works.
o Fixed a couple of bugs in MinGW.
- Fixed unknown type sigset_t error in MinGW.
- Fixed pointer to integer cast warnings in MinGW.

Version 6.6.8 - December 17 2021

[CHANGES]
o gtags: Now you can write comments in '.notfunction' file.
  Lines starting with ';' are comment lines.

	[.notfunction]
	+---------------------------
	|;
	|; This is a comment.
	|;
	|main

o configure: Now you can build Global with external SQLite3 library.
  --with-sqlite3[=dir]    By default, the embedded version is used.

[NEW DOCUMENT]
o plugin-factory/PLUGIN_HOWTO.universal-ctags:
  Added a howto document: 'How to get Universal Ctags to handle references'

[FIXED BUGS]
o global: The --print0 option was ignored when using the --completion command.
  Now it works.
o gtags: The --file option could cause segmentation fault.
  Now it works.

Version 6.6.7 - July 1 2021

[FIXED BUGS]
o Pygments-enabled gtags failed to start. Now it works.
  (Changed the shebang of pygments_parser.py.in from '#!/usr/bin/env python'
  to '#!@PYTHON@'.)

Version 6.6.6 - May 2 2021

[CHANGES]
New facilities:
o gtags.el: Added native support for TRAMP.
  'global-client' command is no longer needed to use TRAMP.

  Required:
    GNU Emacs version 25.1 or later (Tested on version 26).
    If you don't use 'gtags-visit-rootdir' then version 22 is good enough.
    You also need to install Global at the remote site.
  Limitations: gtags-path-style: 'root is automatically converted to 'relative.

o gtags.el: New custom variable 'gtags-path-order'.
  value: 'alphabet (default) or 'nearness.

[DEPRECATED]
o gtags: environment variable GTAGSFORCECPP is deprecated.

[FIXED BUGS]
o configure: configure script failed with an error 'realpath not found.' on MacOS 11.1.
  Now, it works.
o gtags: If GTAGSFORCECPP was defined and --explain was used and the first file is a
  '*.h', gtags crashed. Now it works.

Version 6.6.5 - September 3 2020


gtags.el
;;; gtags.el --- gtags facility for Emacs

;;
;; Copyright (c) 1996, 1997, 1998, 1999
;;            Shigio Yamaguchi. All rights reserved.
;; Copyright (c) 1999
;;            Tama Communications Corporation. All rights reserved.
;;
;; Redistribution and use in source and binary forms, with or without
;; modification, are permitted provided that the following conditions
;; are met:
;; 1. Redistributions of source code must retain the above copyright
;;    notice, this list of conditions and the following disclaimer.
;; 2. Redistributions in binary form must reproduce the above copyright
;;    notice, this list of conditions and the following disclaimer in the
;;    documentation and/or other materials provided with the distribution.
;; 3. All advertising materials mentioning features or use of this software
;;    must display the following acknowledgement:
;;      This product includes software developed by Tama Communications
;;      Corporation and its contributors.
;; 4. Neither the name of the author nor the names of any co-contributors
;;    may be used to endorse or promote products derived from this software
;;    without specific prior written permission.
;; 
;; THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
;; ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
;; IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
;; ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
;; FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
;; DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
;; OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
;; HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
;; LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
;; OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
;; SUCH DAMAGE.
;;
;;	gtags.el	17-Sep-99
;;

;; This file is part of GLOBAL.
;; GLOBAL home page is at: http://www.tamacom.com/global/
;; Author: Shigio Yamaguchi <shigio@tamacom.com>
;; Version: 1.53
;; Keywords: tools

;;; Code

(defvar gtags-buffer-stack nil
  "Stack for tag browsing.")
(defvar gtags-point-stack nil
  "Stack for tag browsing.")
(defvar gtags-complete-list nil
  "Gtags complete list.")
(defvar gtags-history-list nil
  "Gtags history list.")
(defconst symbol-regexp "[A-Za-z_][A-Za-z_0-9]*"
  "Regexp matching tag name.")
(defconst definition-regexp "#[ \t]*define[ \t]+\\|ENTRY(\\|ALTENTRY("
  "Regexp matching tag definition name.")
(defvar gtags-read-only nil
  "Gtags read only mode")
(defvar gtags-mode-map (make-sparse-keymap)
  "Keymap used in gtags mode.")
(defvar running-xemacs (string-match "XEmacs\\|Lucid" emacs-version)
  "Whether we are running XEmacs/Lucid Emacs")
(define-key gtags-mode-map "\et" 'gtags-find-tag)
(define-key gtags-mode-map "\er" 'gtags-find-rtag)
(define-key gtags-mode-map "\es" 'gtags-find-symbol)
(define-key gtags-mode-map "\eg" 'gtags-find-pattern)
(define-key gtags-mode-map "\el" 'gtags-find-file)
(define-key gtags-mode-map "\ec" 'gtags-make-complete-list)
(define-key gtags-mode-map "\C-]" 'gtags-find-tag-from-here)
(define-key gtags-mode-map "\eh" 'gtags-display-browser)
(define-key gtags-mode-map "\C-t" 'gtags-pop-stack)
(define-key gtags-mode-map "\e." 'etags-style-find-tag)
(if (not running-xemacs) nil
 (define-key gtags-mode-map 'button2 'gtags-find-tag-by-event)
 (define-key gtags-mode-map 'button3 'gtags-pop-stack))
(if running-xemacs nil
 (define-key gtags-mode-map [mouse-2] 'gtags-find-tag-by-event)
 (define-key gtags-mode-map [mouse-3] 'gtags-pop-stack))

(defvar gtags-select-mode-map (make-sparse-keymap)
  "Keymap used in gtags select mode.")
(define-key gtags-select-mode-map "q" 'gtags-pop-stack)
(define-key gtags-select-mode-map "\C-t" 'gtags-pop-stack)
(define-key gtags-select-mode-map "\C-m" 'gtags-select-tag)
(define-key gtags-select-mode-map " " 'scroll-up)
(define-key gtags-select-mode-map "\^?" 'scroll-down)
(define-key gtags-select-mode-map "\C-f" 'scroll-up)
(define-key gtags-select-mode-map "\C-b" 'scroll-down)
(define-key gtags-select-mode-map "n" 'next-line)
(define-key gtags-select-mode-map "p" 'previous-line)
(define-key gtags-select-mode-map "j" 'next-line)
(define-key gtags-select-mode-map "k" 'previous-line)
(if (not running-xemacs) nil
 (define-key gtags-select-mode-map 'button2 'gtags-select-tag-by-event)
 (define-key gtags-select-mode-map 'button3 'gtags-pop-stack))
(if running-xemacs nil
 (define-key gtags-select-mode-map [mouse-2] 'gtags-select-tag-by-event)
 (define-key gtags-select-mode-map [mouse-3] 'gtags-pop-stack))

;;
;; utility
;;
(defun util-match-string (n)
  (buffer-substring (match-beginning n) (match-end n)))

;; Return a default tag to search for, based on the text at point.
(defun gtags-current-token ()
  (cond
   ((looking-at "[0-9A-Za-z_]")
    (while (looking-at "[0-9A-Za-z_]")
      (forward-char -1))
    (forward-char 1))
   (t
    (while (looking-at "[ \t]")
      (forward-char 1))))
  (if (and (bolp) (looking-at definition-regexp))
      (goto-char (match-end 0)))
  (if (looking-at symbol-regexp)
      (util-match-string 0) nil))

;; push current context to stack
(defun push-context ()
  (setq gtags-buffer-stack (cons (current-buffer) gtags-buffer-stack))
  (setq gtags-point-stack (cons (point) gtags-point-stack)))

;; pop context from stack
(defun pop-context ()
  (if (not gtags-buffer-stack) nil
    (let (buffer point)
      (setq buffer (car gtags-buffer-stack))
      (setq gtags-buffer-stack (cdr gtags-buffer-stack))
      (setq point (car gtags-point-stack))
      (setq gtags-point-stack (cdr gtags-point-stack))
      (list buffer point))))

;; if the buffer exist in the stack
(defun exist-in-stack (buffer)
  (memq buffer gtags-buffer-stack))

;; is it a function?
(defun is-function ()
  (save-excursion
    (while (and (not (eolp)) (looking-at "[0-9A-Za-z_]"))
      (forward-char 1))
    (while (and (not (eolp)) (looking-at "[ \t]"))
      (forward-char 1))
    (if (looking-at "(") t nil)))

;; is it a definition?
(defun is-definition ()
  (save-excursion
    (if (and (string-match "\.java$" buffer-file-name) (looking-at "[^(]+([^)]*)[ \t]*{"))
	t
      (if (bolp)
	  t
        (forward-word -1)
        (cond
         ((looking-at "define")
	  (forward-char -1)
	  (while (and (not (bolp)) (looking-at "[ \t]"))
	    (forward-char -1))
	  (if (and (bolp) (looking-at "#"))
	      t nil))
         ((looking-at "ENTRY\\|ALTENTRY")
	  (if (bolp) t nil)))))))

;;
;; interactive command
;;
(defun gtags-find-tag ()
  "Input tag name and move to the definition."
  (interactive)
  (let (tagname prompt input)
    (setq tagname (gtags-current-token))
    (if tagname
      (setq prompt (concat ":tag (default " tagname ") "))
     (setq prompt ":tag "))
    (setq input (completing-read prompt gtags-complete-list
                  nil nil nil gtags-history-list))
    (if (not (equal "" input))
      (setq tagname input))
    (push-context)
    (gtags-goto-tag tagname "")))

(defun etags-style-find-tag ()
  "Input tag name and move to the definition.(etags style)"
  (interactive)
  (let (tagname prompt input)
    (setq tagname (gtags-current-token))
    (if tagname
        (setq prompt (concat "Find tag: (default " tagname ") "))
      (setq prompt "Find tag: "))
    (setq input (completing-read prompt gtags-complete-list))
    (if (not (equal "" input)) (setq tagname input))
    (push-context)
    (gtags-goto-tag tagname "")))

(defun gtags-find-symbol ()
  "Input symbol and move to the locations."
  (interactive)
  (let (tagname prompt input)
    (setq tagname (gtags-current-token))
    (if tagname
        (setq prompt (concat "Find symbol: (default " tagname ") "))
      (setq prompt "Find symbol: "))
    (setq input (read-string prompt))
    (if (not (equal "" input)) (setq tagname input))
    (push-context)
    (gtags-goto-tag tagname "s")))

(defun gtags-find-pattern ()
  "Input pattern and move to the locations."
  (interactive)
  (let (tagname prompt input)
    (setq tagname (gtags-current-token))
    (if tagname
        (setq prompt (concat "Find pattern: (default " tagname ") "))
      (setq prompt "Find pattern: "))
    (setq input (read-string prompt))
    (if (not (equal "" input)) (setq tagname input))
    (push-context)
    (gtags-goto-tag tagname "g")))

(defun gtags-find-file ()
  "Input pattern and move to the top of the file."
  (interactive)
  (let (tagname prompt input)
    (setq prompt "Find files: ")
    (setq input (read-string prompt))
    (if (not (equal "" input)) (setq tagname input))
    (push-context)
    (gtags-goto-tag tagname "P")))

(defun gtags-find-rtag ()
  "Input tag name and move to the referenced point."
  (interactive)
  (let (tagname prompt input)
   (setq tagname (gtags-current-token))
   (if tagname
     (setq prompt (concat ":rtag (default " tagname ") "))
    (setq prompt ":rtag "))
   (setq input (completing-read prompt gtags-complete-list
                 nil nil nil gtags-history-list))
   (if (not (equal "" input))
     (setq tagname input))
    (push-context)
    (gtags-goto-tag tagname "r")))

(defun gtags-find-tag-from-here ()
  "Get the expression as a tagname around here and move there."
  (interactive)
  (let (tagname flag)
    (setq tagname (gtags-current-token))
    (if (is-function)
        (if (is-definition) (setq flag "r") (setq flag ""))
      (setq flag "s"))
    (if (not tagname)
        nil
      (push-context)
      (gtags-goto-tag tagname flag))))

(defun gtags-display-browser ()
  "Display current screen on hypertext browser."
  (interactive)
  (let (lno)
    (if (= 0 (count-lines (point-min) (point-max))) nil
    (save-excursion
      (end-of-line)
      (if (equal (point-min) (point))
          (setq lno 1)
        (setq lno (count-lines (point-min) (point)))))
;    (message (number-to-string lno))
    (call-process "gozilla"  nil t nil (concat "+" (number-to-string lno)) buffer-file-name))))

(defun gtags-find-tag-by-event (event)
  "Get the expression as a tagname around here and move there."
  (interactive "e")
  (let (tagname flag)
    (if (= 0 (count-lines (point-min) (point-max)))
        (progn (setq tagname "main") (setq flag ""))
      (if running-xemacs (goto-char (event-point event))
       (select-window (posn-window (event-end event)))
        (set-buffer (window-buffer (posn-window (event-end event))))
        (goto-char (posn-point (event-end event))))
      (setq tagname (gtags-current-token))
      (if (is-function)
          (if (is-definition) (setq flag "r") (setq flag ""))
        (setq flag "s")))
    (if (not tagname)
        nil
      (push-context)
      (gtags-goto-tag tagname flag))))

(defun gtags-select-tag ()
  "Select a tagname in [GTAGS SELECT MODE] and move there."
  (interactive)
  (push-context)
  (gtags-select-it nil))

(defun gtags-select-tag-by-event (event)
  "Select a tagname in [GTAGS SELECT MODE] and move there."
  (interactive "e")
  (if running-xemacs (goto-char (event-point event))
    (select-window (posn-window (event-end event)))
    (set-buffer (window-buffer (posn-window (event-end event))))
    (goto-char (posn-point (event-end event))))
  (push-context)
  (gtags-select-it nil))

(defun gtags-pop-stack ()
  "Move to previous point on the stack."
  (interactive)
  (let (delete context buffer)
    (if (not (exist-in-stack (current-buffer)))
	(setq delete t))
    (setq context (pop-context))
    (if (not context)
	(message "The tags stack is empty.")
      (if delete
	  (kill-buffer (current-buffer)))
      (switch-to-buffer (nth 0 context))
      (goto-char (nth 1 context)))))

;;
;; common function
;;

;; goto tag's point
(defun gtags-goto-tag (tagname flag)
  (let (save prefix buffer lines)
    (setq save (current-buffer))
    (cond
     ((equal flag "P")
      (setq prefix "(P)"))
     ((equal flag "g")
      (setq prefix "(G)"))
     ((equal flag "s")
      (setq prefix "(S)"))
     ((equal flag "r")
      (setq prefix "(R)"))
     (t (setq prefix "(D)")))
    ;; load tag
    (setq buffer (generate-new-buffer (generate-new-buffer-name (concat prefix tagname))))
    (set-buffer buffer)
    (if (equal flag "g") (message "Searching...%s" tagname))
    (if (not (= 0 (call-process "global" nil t nil (concat "-ax" flag) tagname)))
	(progn (message (buffer-substring (point-min)(1- (point-max))))
               (pop-context))
      (goto-char (point-min))
      (setq lines (count-lines (point-min) (point-max)))
      (cond
       ((= 0 lines)
         (cond
          ((equal flag "P")
           (message "%s: path not found" tagname))
          ((equal flag "g")
           (message "%s: pattern not found" tagname))
          ((equal flag "s")
           (message "%s: symbol not found" tagname))
          (t
           (message "%s: tag not found" tagname)))
	(pop-context)
	(kill-buffer buffer)
	(set-buffer save))
       ((= 1 lines)
	(gtags-select-it t))
       (t
	(switch-to-buffer buffer)
	(gtags-select-mode))))))

;; select a tag line from lines
(defun gtags-select-it (delete)
  (let (line file)
    ;; get context from current tag line
    (beginning-of-line)
;;    (if (not (looking-at "[A-Za-z_][A-Za-z_0-9]*[ \t]+\\([0-9]+\\)[ \t]\\([^ \t]+\\)[ \t]"))
    (if (not (looking-at "[^ \t]+[ \t]+\\([0-9]+\\)[ \t]\\([^ \t]+\\)[ \t]"))
        (pop-context)
      (setq line (string-to-number (util-match-string 1)))
      (setq file (util-match-string 2))
      (if delete (kill-buffer (current-buffer)))
      ;; move to the context
      (if gtags-read-only (find-file-read-only file) (find-file file))
      (goto-line line)
      (use-local-map gtags-mode-map))))

;; make complete list
(defun gtags-make-complete-list ()
  "Make tag name list for completion."
  (interactive)
  (save-excursion
    (message "Making completion list...")
    (setq gtags-complete-list (make-vector 63 0))
    (set-buffer (generate-new-buffer "*Completions*"))
    (call-process "global" nil t nil "-c")
    (goto-char (point-min))
    (while (looking-at symbol-regexp)
      (intern (util-match-string 0) gtags-complete-list)
      (forward-line))
    (message "Making completion list... done")
    (kill-buffer (current-buffer))))

;;;###autoload
(defun gtags-mode ()
  "Minor mode for browsing C source using GLOBAL."
  (interactive)
  (use-local-map gtags-mode-map)
  (run-hooks 'gtags-mode-hook))

;; make gtags select mode
(defun gtags-select-mode ()
  "Major mode for choosing a tag from tags list."
  (setq buffer-read-only t
        major-mode 'gtags-select-mode
        mode-name "Gtags Select")
  (use-local-map gtags-select-mode-map)
  (setq truncate-lines t)
  (goto-char (point-min))
  (message "[GTAGS SELECT MODE] %d lines" (count-lines (point-min) (point-max)))
  (run-hooks 'gtags-select-mode-hook))

;;; gtags.el ends here
